dvc/cache/local.py

```diff
@@ -437,14 +437,26 @@ def push(self, named_cache, remote, jobs=None, show_checksums=False):
 
     @index_locked
     def pull(self, named_cache, remote, jobs=None, show_checksums=False):
-        return self._process(
+        ret = self._process(
             named_cache,
             remote,
             jobs=jobs,
             show_checksums=show_checksums,
             download=True,
         )
 
+        if not remote.tree.verify:
+            for checksum in named_cache.scheme_keys("local"):
+                cache_file = self.tree.hash_to_path_info(checksum)
+                if self.tree.exists(cache_file):
+                    # We can safely save here, as existing corrupted files will
+                    # be removed upon status, while files corrupted during
+                    # download will not be moved from tmp_file
+                    # (see `BaseTree.download()`)
+                    self.tree.state.save(cache_file, checksum)
+
+        return ret
+
```

Comment:
More natural to have it here and when it is here, it is obvious that we should use `self.tree.state` and not `self.repo.state`